dataikuapi/dss/ml.py

```diff
@@ -915,53 +933,78 @@ class PredictionAlgorithmSettings(dict):
     """
     Object to read and modify the settings of a prediction ML algorithm.
 
-    Do not create this object directly, use :meth:`DSSMLTask.get_algorithm_settings(algorithm)` instead
+    Do not create this object directly, use :meth:`DSSMLTask.get_algorithm_settings` instead
     """
 
     def __init__(self, raw_settings, hyperparameter_search_params):
         super(PredictionAlgorithmSettings, self).__init__(raw_settings)
         self._hyperparameter_search_params = hyperparameter_search_params
         self._hyperparameters_registry = dict()
+        self._attr_to_json_remapping = dict()
 
-    def __setattr__(self, key, value):
-        if not hasattr(self, key):
+    def __setattr__(self, attr_name, value):
+        if not hasattr(self, attr_name):
             # call from __init__
-            super(PredictionAlgorithmSettings, self).__setattr__(key, value)
-        elif key in self._hyperparameters_registry:
-            # syntactic sugars
-            target = self._hyperparameters_registry[key]
-            if isinstance(target, (SingleValueHyperparameterSettings, SingleCategoryHyperparameterSettings)):
-                target.set_value(value)
-            elif isinstance(target, CategoricalHyperparameterSettings):
-                target.set_values(value)
-            else:
-                raise Exception("Invalid assignment of a NumericalHyperparameterSettings object")
-        elif key == "lambda_":
-            raise Exception("Invalid assignment of a NumericalHyperparameterSettings object")
+            super(PredictionAlgorithmSettings, self).__setattr__(attr_name, value)
         else:
-            # other cases (properties setter, new attribute...)
-            super(PredictionAlgorithmSettings, self).__setattr__(key, value)
+            if attr_name in self._attr_to_json_remapping:
+                # attribute name and json key mismatch (e.g. "lambda", "alphaMode")
+                attr_name = self._attr_to_json_remapping[attr_name]
+            if attr_name in self._hyperparameters_registry:
+                # syntactic sugars
+                target = self._hyperparameters_registry[attr_name]
+                if isinstance(target, (SingleValueHyperparameterSettings, SingleCategoryHyperparameterSettings)):
+                    target.set_value(value)
+                elif isinstance(target, CategoricalHyperparameterSettings):
+                    target.set_values(value)
+                elif isinstance(target, NumericalHyperparameterSettings):
+                    raise Exception("Invalid assignment of a NumericalHyperparameterSettings object")
+                else:
+                    # simple parameter
+                    assert isinstance(value, type(target)), "Invalid type {} for parameter {}: expected {}".format(type(value), attr_name, type(target))
+                    super(PredictionAlgorithmSettings, self).__setattr__(attr_name, value)
+                    self[attr_name] = value
```

Comment:
Shouldn't the first line suffice ?

Reply:
Actually no, e.g. for attribute `epsilon`:
- `super(PredictionAlgorithmSettings, self).__setattr__(attr_name, value)` handles the attribute `epsilon` of the `PredictionAlgorithmSettings` object
- `self[attr_name] = value` handles the dict value at key `"epsilon"` which is different from the attribute

Reply:
Yes, you're right, I got confused my bad

Reply:
Maybe it warrants a comment to make sure nobody accidentally removes that second line?